Add the i801 SMBus driver

Steve-Tech opened the pull request:
Hi, I've ported the i801 I2C/SMBus driver from the Linux kernel. I've also added a `BIT` macro to `core.inc`.

This protocol is commonly used on Intel motherboards, and could allow querying of RAM temperature, and configuring RGB.

Only the read byte/word has been properly tested, since I didn't know what I could write to on my system.
This was also developed on a decade old system, so hopefully it works on newer ones.

For myself and everyone's reference:
- The best Intel chipset datasheet I could find: https://www.intel.com/content/dam/www/public/us/en/documents/datasheets/9-series-chipset-pch-datasheet.pdf
- The newest usable Intel chipset datasheet I could find: https://edc.intel.com/content/www/us/en/design/products-and-solutions/processors-and-chipsets/comet-lake-u/intel-400-series-chipset-on-package-platform-controller-hub-register-database/introduction/
- The code this is based off: https://github.com/torvalds/linux/blob/v6.14/drivers/i2c/busses/i2c-i801.c

Thanks,
Steve

namazso commented:
Thanks, I'll try properly reviewing this against the datasheet this weekend. A quick preliminary one though: You have some variables named `ret`. If they hold a status (which they do), recommended name would be `status` and checking them with `NT_SUCCESS` even if you're fairly sure what values it could take. NTSTATUS has non-failure codes, see [MSDN](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/87fba13e-bf06-450e-83b1-9241dc81e781). PawnIO borrows heavily from NT development practices, for example by having all natives return NTSTATUSes as well.

Steve-Tech commented:
Thanks for the preliminary feedback! I've just renamed all the variables that hold an NTSTATUS to `status`, and checked them with `NT_SUCCESS`. Some of the existing `status` variables actually held a value obtained from the HSTSTS register, so I've renamed them to `hststs` for readability.

I also renamed the module to SmbusI801 because I think it better describes what it does.
And I also made some of the same changes to my PIIX4 driver too.

hexawyz commented:
Hey, thanks a lot for working on this. I was starting to motivate myself to work on such a driver, as I've finished my current side-quest, but it seems I won't have to. I will glady take some time to test this driver to validate that it works. (And I do have some code making use of block writes)

In the mean time, I took I quick look at the code and it seems to me that the block commands would require splitting every byte in its own "cell" instead of accepting data in a more raw form. If that's the case, it would incurr significant overhead to call these methods. (x8 memory + cpu cost of doing the transformation and copies)

It would be preferable if the block methods accepted/outputed a parameter that is the length in bytes, followed by raw byte data packed into the cells. The little endian architecture would still make this easy to process, and it would be more efficient overall.

For example:

````
ioctl_i801_write_block_data
IN = 
[0] 0x00000000000000AA // Address
[1] 0x00000000000000CC // Command / Register
[2] 0x0000000000000020 // Length
[3] 0x0706050403020100 // Data bytes 0
[4] 0x0F0E0D0C0B0A0908 // Data bytes 1
[5] 0x1716151413121110 // Data bytes 2
[6] 0x1F1E1D1C1B1A1918 // Data bytes 3
````

Here, instead of transmitting 272 = (8 + 8 + 32 * 8) bytes, we'd transmit only 48 = (8 + 8 + 8 + 4*8), which would be a significant improvement (5.6x less, and everything would fit in a single cache line)

I've also seen that you do create intermediate copies of data before calling your internal function. It might be preferable to just pass the input array and an offset (+ length ?) to the actual data, especially in the case of block operations. Zero-copy is almost always better and I think it would be the case here too.

Since I haven't looked at the AMD-version of your driver, I can't say much, but I would assume those remarks apply there too

namazso commented:
@hexawyz Just a FYI about zero copy: You do need to be careful with the ioctl input/output buffers because in a windows ioctl they actually share memory, and PawnIO just maps them straight into the virtual address space. That is, if you write out[4], the contents of in[0] will change (the 32 byte difference is due to ioctl name).

Steve-Tech commented:
@hexawyz I did actually think of doing zero-copies originally, but I didn't want to do something too extravagant if I had no way of testing it.

Anyway, I've implemented both zero-copy and packing block transactions into 4 cells.

Although I have I feeling I've screwed up the endianness for the last one, like the first read/written byte, should be the least significant for little endian right?

namazso commented:
I reverted the last two commits, the packing was buggy either way (forgot to multiply by 8). I reimplemented the packing with some extra copying as a stock (as it's not like we care a lot about performance here, this thing is all interpreted, and also moved most things to static array sizes. The Pawn compiler inserts automatic range checks for those to catch invalid access.

Would be great if you could test if the module still works as intended, the interface should be the same as with your latest version, except that you need to always allocate all 32 bytes of block data.

namazso commented:
> It being interpreted makes every operation cost even more, so we should try to be lean when possible. (Although performance matters all the time IMO)

It's still within the nanosecond-microsecond range, probably negligible even compared to the cost of `in` / `out` instructions. On the other hand, if you get unlucky, you'll get descheduled by the scheduler for up to 16ms mid-operation. If you truly want low latency, you'd want to disable interrupts during operations, preventing that from happening. The problem is that that can mess up the latency of everything else, and I'd argue that RGB is less important than not having stuttering in games.

Additionally, in user mode (where whatever RGB control software runs) you can't really fight the scheduler any more than raising priority and hoping, so the RAM stick desync would occasionally happen anyway! If you wanted it to be low latency during the whole set of operations, you'd implement the whole thing in Pawn with interrupts disabled. However I consider that way overkill and much harder to audit even if we discount the issue of messing up latency for everyone else.

hexawyz commented:
I'm not saying that we should shoot for real-time guarantees though.
Just that we should try to minimize the cost so that we are good in the optimistic scenario.

And it's not only about delay too. Every CPU cycle that we consume could have been used by something else, or preferably not at all.

namazso commented:
It really doesn't matter when it's so negligible, especially compared to the operation itself we're doing. But even if it did - unpacking the bytes in Pawn costs much more than in native usermode code, so you'd want to **not** pack things instead. I only went with packing for convenience's sake for the user, so that a normal `u8[32]` can be just thrown into the struct instead of the weirdness of cells.

> Just that we should try to minimize the cost so that we are good in the optimistic scenario.

Performance just **shouldn't** be the concern of a Pawn module, because it just shouldn't run enough code for it to matter. Nothing here was designed with that view. Just to list a few things:

- PawnPP doesn't implement anything outside the core instruction set. (such as byte packing instructions)
- The used PawnPP memory model range checks all accesses.
- 64-bit cells were picked for simplicity of APIs. Since Pawn instructions are also per cell, that means we waste about 90% of the memory we allocate for the code.
- PawnIO doesn't implement anything from the Pawn standard library (we don't even have `min` and `max`, let alone floats) as they should be safer to implement in Pawn.
- PawnIO only uses wide strings (every character is a cell) everywhere.
- All PawnIO modules are compiled with bounds checks enabled.
- PawnIO modules are single threaded, even though the VM design leaves enough liberty to implement a SMT model.
- We use `new` everywhere which zeroes cells. If you ever wrote AMXXPawn / SourcePawn plugins you know that `static` is preferable for speed (which actually matters there) because it doesn't zero.

And there's still the point that PawnIO will always be lightyears ahead of WinRing0 performance (which so far was seemingly acceptable) simply because a syscall costs a stupid amount of compute compared to the extremely efficient interpreter code the compiler made.

So yeah, all in all, just don't worry about micro optimizations. If you feel the need to worry about performance, your code does too much.

hexawyz commented:
I did the change here. Code builds, but I haven't add the time to setup a test environment yet:
https://github.com/Steve-Tech/PawnIO.Modules/compare/I2cI801...hexawyz:PawnIO.Modules:I2cI801

hexawyz commented:
OK, so far I tested (using the commit I linekd above)

* `ioctl_i801_write_quick`
* `ioctl_i801_write_byte`
* `ioctl_i801_write_byte_data`
* `ioctl_i801_read_byte_data`
* `ioctl_i801_write_word_data`

These seem to work fine.
I currently don't have any use for any of the other methods, but I think I should be able to at least test the word read and block read scenarios.
I also need to do a few more tests on the above methods, but so far, everything seems good.

hexawyz commented:
I have now additionally tested:

* `ioctl_i801_read_word_data`
* `ioctl_i801_read_block_data`

Note that the `io_in_bytes` function i showed above was returning bytes in the reverse order, but I fixed it in my last commit, so the code here is up-to-date with what I tested: https://github.com/hexawyz/PawnIO.Modules/tree/a49ae988f439355535d31a9efcc372ab6994ff62

hexawyz commented:
@Steve-Tech I rebased my branch on your latest changes and things seems to still be working. (I didn't do a in-depth test, but my app starts, which is good enough)

Also I did a cold boot test earlier, which was the final test I needed to do.
The RGB RAM sticks need to be moved around upon a cold start (AC off), which involves slightly different code paths than the "hot" state, and everything went fine. I consider this to be working good on my side.

BTW, I tested the block read method with the following sizes: 3, 4, 24. Got an error when requesting 32 but I would assume that's a problem on the device side. (I have no idea if it is supposed to support this at all)
Block write was tested with only 24 I think, which did work. (Small caveat is that this is a multiple of 8, but I would be surprised if it worked any different for another size)

namazso commented on SmbusI801.p:
```diff
@@ -0,0 +1,857 @@
+            goto cleanup;
+        }
+
+        if (len > I2C_SMBUS_BLOCK_MAX)
```
True, I didn't notice that also had a check. But with static array size, the compiler will also insert a bounds check anyway.

namazso commented on SmbusI801.p:
```diff
@@ -0,0 +1,857 @@
+    if (out_size < 1)
+        return STATUS_BUFFER_TOO_SMALL;
+
+    new address = in[0];
```
Separate variable helps readability, since IDE support is not quite there for Pawn.

> By my tests, such a small change reduces the size of the file by 72 bytes (?!)

Yes, each Pawn instruction is 8 bytes (16 if they have argument) because they're cell sized.

namazso commented:
Signed copy: [SmbusI801.zip](https://github.com/user-attachments/files/19934490/SmbusI801.zip)

Btw these are covered by the same LGPL 2.1 and are redistributable. Since it's "dynamic linked" all you need to comply with is the notice and providing / shipping the source too. Nothing infectious. The driver takes blobs so that you can just embed these in your software, complying with just the notice requirement.

namazso commented:
Something I didn't notice with this is that there was still quite some mixup between NTSTATUS and hststs values, this came to light as I added tagging for a few types. I attempted fixing it in https://github.com/namazso/PawnIO.Modules/commit/21d0deb8be4f308293f17be20d6b17f0092a8e80

Steve-Tech commented:
Yeah sorry, I didn't wanna stray too far from the Linux driver, but I should've asked if those occurrences were okay.

I've tested the following ioctls and they seem fine:
- `ioctl_i801_read_byte`
- `ioctl_i801_read_byte_data`
- `ioctl_i801_read_word_data`
